### ISSUE TITLE
Implement BaseDetector Pipeline and EPR/WEPR aliases

### ISSUE THREAD

**Issue by chicham** [artefactory/artefactual#240]
Implement the `BaseDetector` pipeline class to replace the standalone legacy detectors.

## Implementation decision
Instead of writing bespoke logic for each reduction (EPR / WEPR), `BaseDetector` inherits directly from `sklearn.pipeline.Pipeline`. This standardizes the entire feature extraction and classification process. 
- Sequence-level classification is handled automatically by the native `.predict_proba()` method inherited from `Pipeline` (we do **not** override it — design decision #2).
- Token-level classification is handled by a custom `.predict_token_proba()` method.

## Construction (two paths)
- `EPR(**kwargs)` / `WEPR(**kwargs)` build an **unfitted** detector (a fresh `PretrainedLogisticRegression`) — for calibration training via `.fit()` (#282). Calling `predict*` before fitting raises `NotFitted`.
- `BaseDetector.from_pretrained(model, reduction)` builds a **ready-to-predict** detector with weights loaded (#279).

Both share the same step structure; the transformers are stateless (`requires_fit = False`, #277/#278), so only the classifier carries fitted state.

## Two prediction paths
Both methods share the parser step, then diverge at `EntropyTransformer`: `predict_proba` reduces over tokens (one score per sequence), `predict_token_proba` keeps one score per token.

```mermaid
flowchart TD
    X["X — raw outputs"] --> P["LogProbParser.transform<br/>(n, max_tokens, k)"]
    subgraph seq["predict_proba — sequence level"]
        direction TB
        SA["EntropyTransformer.transform<br/>reduce over tokens<br/>(n, n_features)"] --> SC["classifier.predict_proba<br/>→ (n, 2); [:, 1] = P(halluc.)"]
    end
    subgraph tok["predict_token_proba — token level"]
        direction TB
        TA["EntropyTransformer.transform_tokens<br/>per token, no reduction<br/>(n, max_tokens, n_features)"] --> TC["flatten → classifier.predict_proba → reshape<br/>→ (n, max_tokens, 1) one score per token"]
    end
    P --> SA
    P --> TA
```

## Algorithm for predict_token_proba
`predict_token_proba(X)` manually routes data through the pipeline in token mode, then **masks padded tokens before the classifier** (which rejects NaN):
1. **Apply transformations (token mode):** Iterate `self._iter(with_final=False)`. For each transformer, call `transform_tokens(Xt)` if present, else fall back to `transform(Xt)`. The output is a padded 3D array `(n_samples, max_tokens, n_features)` with **`NaN` at padded token positions**.
2. **Flatten:** reshape to `(n_samples * max_tokens, n_features)`.
3. **Mask padded rows:** `valid = ~np.isnan(rows).any(axis=1)`. This is required — `LogisticRegression.predict_proba` runs `check_array(ensure_all_finite=True)` and raises `ValueError` on any `NaN` row, which would otherwise hit on every variable-length batch.
4. **Predict on valid rows only:** `classifier = self.steps[-1][1]`; `p = classifier.predict_proba(rows[valid])[:, 1]` (positive class = hallucination).
5. **Scatter + reshape:** prefill a `(n_samples * max_tokens,)` output with `NaN`, set `out[valid] = p`, reshape to `(n_samples, max_tokens, 1)`. Padded positions stay `NaN` (there is no token there).

## Empty sequences
A completion with zero tokens produces an all-`NaN` feature row. Since `predict_proba` is the inherited (non-overridden) `Pipeline.predict_proba`, the NaN must be resolved **upstream**: `EntropyTransformer.transform` (#277) maps an all-`NaN` (token-less) sequence to a **zero feature vector**, so the classifier returns the calibrated baseline `sigmoid(intercept_)` (matches legacy). For `predict_token_proba`, an empty sequence's token rows are all masked out → that sequence's tokens are all `NaN` in the output.

## Skeleton (signatures & contracts — bodies to be implemented)
```python
from sklearn.pipeline import Pipeline
import numpy as np

class BaseDetector(Pipeline):
    def predict_token_proba(self, X) -> np.ndarray:
        """Token-level hallucination probabilities, shape (n_samples, max_tokens, 1).

        Route the pipeline in token mode (transform_tokens, with transform fallback),
        mask padded (NaN) token rows before the classifier, scatter results back so
        padded positions stay NaN. See "Algorithm for predict_token_proba".
        """
        raise NotImplementedError

    @classmethod
    def from_pretrained(cls, pretrained_model_name_or_path: str, reduction: str) -> "BaseDetector":
        """Build a ready-to-predict detector with weights loaded (#279). See acceptance criteria."""
        raise NotImplementedError


# Factory aliases (module level, exposed via #237) — each returns a BaseDetector wired as
#   LogProbParser -> EntropyTransformer(reduction="epr"|"wepr") -> PretrainedLogisticRegression
def EPR(**kwargs) -> "BaseDetector":
    raise NotImplementedError

def WEPR(**kwargs) -> "BaseDetector":
    raise NotImplementedError
```

## Acceptance criteria
- [x] Inherits directly from `sklearn.pipeline.Pipeline`.
- [x] Factory functions `EPR(**kwargs)` and `WEPR(**kwargs)` return a `BaseDetector` with steps `LogProbParser` → `EntropyTransformer(reduction="epr")` / `EntropyTransformer(reduction="wepr")` → `PretrainedLogisticRegression` (the `reduction` keyword matches #277; classifier per #279).
- [x] `predict_token_proba(X)` routes in token mode and **masks padded (`NaN`) token rows** before `classifier.predict_proba`, scattering results back so padded positions stay `NaN`; output `(n_samples, max_tokens, 1)`. Verified on a variable-length batch (different sequence lengths).
- [x] Empty (token-less) sequences are handled: `predict_proba` returns the baseline (via the zero-feature mapping in #277), `predict_token_proba` returns all-`NaN` tokens for that sequence.
- [x] `from_pretrained(cls, pretrained_model_name_or_path, reduction)` builds and returns a ready-to-predict `BaseDetector` — steps `LogProbParser`, `EntropyTransformer(reduction=...)`, `PretrainedLogisticRegression.from_pretrained(...)` (#279) — propagating loaded hyperparameters (e.g. `k`). No `fit` call is needed: the transformers are stateless (`requires_fit = False`, #277/#278) and the classifier is pre-fitted.

**Comment by CharlotteLeBihan18** [comment]
there should only be one return and it is the classifier that should be using from_pretrained
